Add Extensions API blog post for documentId and friends. (#3903)
GoogleChrome/developer.chrome.com: site/en/blog/extension-instantnav/index.md

---
layout: 'layouts/blog-post.njk'
title: "Chrome Extensions: Extending API to support Instant Navigation"
description: >
    The Extensions API has been updated to support back/forward cache, preloading navigations.
authors:
  - dtapuska
hero: 'image/zKrSUSkPboWMTTSEkowJbqw5Egi2/r5IdmaTdE5wBS2d6CCQY.png'
alt: >
  Instant navigation in a tab.
date: 2022-10-04
---

TL;DR: You may need to update your extension if it assumes `frameId == 0`

> is a page’s main frame. See below for details.

Reviewer: There is a lot more in this article than `frameId`.  The text from the description would be better here.

Author: Simeon wrote the TLDR, but changed to description I wrote.

Chrome has been working hard at making navigation fast. Instant Navigation
technologies such as [Back/Forward Cache](https://web.dev/bfcache/)
([shipped](https://chromestatus.com/feature/6279906713403392) on desktop in M96)
and [Speculation Rules](https://chromestatus.com/feature/5740655424831488)
(shipped in M103) improve both the going back and going forward experience.
In this post we will explore the updates we’ve made to browser extensions APIs
to accommodate these new workflows.

> ## Understanding the Types of Pages

Reviewer: ```suggestion
## Understanding the types of pages
```

Reviewer: This site uses sentence case for titles. Please fix this throughout. (I know it's not what your writing teachers taught you.)

Author: Done

Prior to the introduction of Back/Forward Cache and Prerendering, an individual
tab only had one active page. This was always the one that was visible. If a

> tab was navigated backwards, the active page would be destroyed (Page B) and

Reviewer: >If a tab was navigated backwards...

Why not 'if a user returns to the previous page...'?

Author: Done

{% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/VR5hJLKud7lNIou8baHx.png", alt="Eviction of active page", width="350", height="209" %}

With Back Forward Cache and Prerendering, there is no longer a 1:1 relationship
between tabs and pages. Now, each tab actually stores multiple pages and pages
transition between states rather than being destroyed and reconstructed.

For example, a page could begin its life as a prerendered (not visible) page,
transition into an active (visible) page when the user clicks a link, and then
be stored in the Back Forward Cache (not visible) when the user navigates to
another page, all without the page ever being destroyed. Later in this article
we will look at the new properties exposed to help extensions understand what
state pages are in.

{% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/0DT7Tzx0vQhNZgegsoEj.png", alt="Types of Pages", width="531", height="209" %}

Note that a tab can have a series of prerender pages (not just one), a single
*active* (visible) page, and a series of Back/Forward cached pages.

## What’s changing for extension developers?
### FrameId == 0
In Chromium, we refer to the topmost/main frame as the outermost frame.

> Extension authors that assume the [*frameId*](/docs/extensions/reference/webNavigation/#a-note-about-frame-ids:~:text=onErrorOccurred%20event%20fired.-,frameId,-number)
> of the outermost [frame is 0](/docs/extensions/reference/webNavigation/#a-note-about-frame-ids)
> may have problems. Since a tab can now have multiple outermost frames
> (prerendered and cached pages), the assumption that there is a single outermost
> frame for a Tab is incorrect. frameId == 0 will still continue to represent the
> outermost frame of the **active** page, but the outermost frames of **other**
> pages will be non-zero. A new field frameType has been added to fix this
> problem. See the [“How do I determine if a frame is the outermost frame?”](#outermost-frame)
> section.

Reviewer 2: @dtapuska, as you likely know based on the use of your text fragment link, there isn't a great way to link directly to the *frameId* property on the callback of `webNavigation.getAllFrames()`. There also isn't a great description of what this property is or how it works today. I think it may be best if we expand the reference documentation to explicitly describe the roles of *frameId* and *documentId*. 

```suggestion
Extension authors that assume the
[*frameId*]((/docs/extensions/reference/webNavigation/#:~:text=onErrorOccurred%20event%20fired.-,frameId,-number))
of the outermost frame is 0 (a previous best practice) may have problems. Since
a tab can now have multiple outermost frames (prerendered and cached pages), the
assumption that there is a single outermost frame for a Tab is incorrect.
`frameId == 0` will still continue to represent the outermost frame of the
**active** page, but the outermost frames of **other** pages in the same tab
will be non-zero. A new field *frameType* has been added to fix this problem.
See the [How do I determine if a frame is the outermost
frame?](#outermost-frame) section of this post.
```

I want to try to steer away from linking readers to the [A note about frame IDs](/docs/extensions/reference/webNavigation/#a-note-about-frame-id) section. It's not a great look to point to the live version of our own docs to highlight of date guidance. If we _do_ want to direct readers to out of date guidance, I'd be more comfortable linking to a Wayback Machine copy rather than the live site. 

I'd very much like to update the webNavigation API's guidance either in this PR or in another that follows shortly after. @dtapuska and @jpmedley, do you have a preference on which path we take? Given how long Dave has been waiting on this PR, I'm inclined to tackle it separately but I'm open to either.

Reviewer: I'll take care of the API guidance. Let's not wait for it.

lifecycle it is in. See [“How do I tell what life cycle the page is in?”](#lifecyle)
and [“How do I determine when a page transitions?”](#transitions) sections.

## How do I tell what life cycle the page is in? {: #lifecyle}

The [documentLifecycle](/docs/extensions/reference/extensionTypes/#type-DocumentLifecycle)
attribute has been added to a number of extensions APIs where the *frameId* was
previously available. If the value is present on an event (such as
[onCommitted](/docs/extensions/reference/webNavigation/#event-onCommitted)),
the value is the state at which the event was generated. You can always query
information from the [WebNavigation GetFrame/GetAllFrames](/docs/extensions/reference/webNavigation/#method-getFrame)
APIs, but using the value from the event is always preferred. If you do use the
[GetFrame](/docs/extensions/reference/webNavigation/#method-getFrame) API be
aware the state of the frame may have changed from when the event was generated
to when the GetFrame promise is resolved.

The [documentLifecycle](/docs/extensions/reference/extensionTypes/#type-DocumentLifecycle)
has the following values:

> - **prerender** : Not currently presented to the user but preparing to possibly be displayed to the user.
> - **active**: Presently displayed to the user.
> - **cached**: Stored in the back/forward cache.
> - **pending_deletion**:  A special case where the document is actively being destroyed.

Reviewer 2: Update formatting to better indicate these are strings that will be returned by the browser.

```suggestion
- `"prerender"`: Not currently presented to the user but preparing to possibly
  be displayed to the user.
- `"active"`: Presently displayed to the user.
- `"cached"`: Stored in the back/forward cache.
- `"pending_deletion"`:  A special case where the document is actively being
  destroyed.
```

## How do I determine if a frame is the outermost frame? {: #outermost-frame}

> Previously extensions may have used checking [*frameId*](/docs/extensions/reference/webNavigation/#a-note-about-frame-ids)
> == 0 to determine if the event occurring is for the outermost frame or not.
> With multiple pages in a tab we now have multiple outermost frames, so the
> definition of *frameId* is problematic. You will never receive events about a
> Back/Forward cached frame. However, for prerender frames the *frameId* will be
> non-zero for the outermost frame. So using *frameId* == 0 as a signal for
> determining if it is the outermost frame is incorrect.

Reviewer 2: ```suggestion
Previously extensions may have checked whether `frameId == 0` to determine if
the event occurring is for the outermost frame or not. With multiple pages in a
tab we now have multiple outermost frames, so the definition of *frameId* is
problematic. You will never receive events about a Back/Forward cached frame.
However, for prerender frames the *frameId* will be non-zero for the outermost
frame. So using `frameId == 0` as a signal for determining if it is the
outermost frame is incorrect.
```

To help with this, we introduced a new field called [*frameType*](/docs/extensions/reference/extensionTypes/#type-FrameType)
so determining if the frame is indeed the outermost frame is now easy.

The [*frameType*](/docs/extensions/reference/extensionTypes/#type-FrameType)
has the following values:

> - **outermost_frame**: Typically referred to the topmost frame. Note that there are multiple of
> these: there could be many for the various documentLifecycles (such as a
> prerendered outermost_frame and a cached outermost_frame).
> - **fenced_frame**: A special type of frame that is embedded inside other frames
> (see [fenced-frames](https://github.com/WICG/fenced-frame))
> - **sub_frame**: Typically an iframe

Reviewer 2: Update formatting to better indicate these are strings that will be returned by the browser.

```suggestion
- `"outermost_frame"`: Typically referred to the topmost frame. Note that there
are multiple of these: there could be many for the various documentLifecycles
(such as a prerendered outermost_frame and a cached outermost_frame).
- `"fenced_frame"`: A special type of frame that is embedded inside other frames
(see [fenced-frames](https://github.com/WICG/fenced-frame))
- `"sub_frame"`: Typically an iframe
```

> We can combine *documentLifecycle* with *frameType* and determine if a frame
> is the active outermost frame. ie.
> (documentLifecycle == “active” && frameType == “outermost_frame”)

Reviewer 2: Prefer "for example" over latin abbreviations. Use monospace formatting for code snippets.

```suggestion
We can combine *documentLifecycle* with *frameType* and determine if a frame is
the active outermost frame. For example, `tab.documentLifecycle == "active"
&& tab.frameType == "outermost_frame"`.
```

## How do I solve time of use problems with frames?

> As we said above a frame hosts a document and the frame may navigate to a new
> document, but the *frameId* will not change. This creates problems when you
> receive an event with just a *frameId*. If you [look up the URL](/docs/extensions/reference/webNavigation/#method-getFrame)
> of the frame it might be different than when the event occured, this is called
> a Time of Use issue. We introduced [*documentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=retrieve%20information%20about.-,documentId,-string%C2%A0optional)
> (and [*parentDocumentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=parentDocumentId))
> to address this issue. The [GetFrame](/docs/extensions/reference/webNavigation/#method-getFrame)
> API now makes the *frameId* optional if a *documentId* is provided. The
> *documentId* will change whenever a frame is navigated.

Reviewer 2: ```suggestion
As we said above a frame hosts a document and the frame may navigate to a new
document, but the *frameId* will not change. This creates problems when you
receive an event with just a *frameId*. If you [look up the
URL](/docs/extensions/reference/webNavigation/#method-getFrame) of the frame it
might be different than when the event occurred, this is called a Time of Use
issue.

To address this, we introduced
[*documentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=retrieve%20information%20about.-,documentId,-string%C2%A0optional)
(and
[*parentDocumentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=parentDocumentId)).
The
[webNavigation.getFrame()](/docs/extensions/reference/webNavigation/#method-getFrame)
method now makes the *frameId* optional if a *documentId* is provided. The
*documentId* will change whenever a frame is navigated.
```

---
layout: 'layouts/blog-post.njk'
title: "Chrome Extensions: Extending API to support Instant Navigation"
description: >
    The Extensions API has been updated to support back/forward cache, preloading navigations.
authors:
  - dtapuska
hero: 'image/zKrSUSkPboWMTTSEkowJbqw5Egi2/r5IdmaTdE5wBS2d6CCQY.png'
alt: >
  Instant navigation in a tab.
date: 2022-11-11
---

TL;DR: The Extensions API has been updated to support back/forward cache,
preloading navigations. See below for details.

Chrome has been working hard at making navigation fast. Instant Navigation
technologies such as [Back/Forward Cache](https://web.dev/bfcache/)

> ([shipped](https://chromestatus.com/feature/6279906713403392) on desktop in M96)

Author: fixed.

and [Speculation Rules](https://chromestatus.com/feature/5740655424831488)

> (shipped in M103) improve both the going back and going forward experience.

Author: fixed

In this post we will explore the updates we’ve made to browser extensions APIs
to accommodate these new workflows.

## Understanding the types of pages

> Prior to the introduction of Back/Forward Cache and Prerendering, an individual

Author: fixed

tab only had one active page. This was always the one that was visible. If a
user returns to the previous page, the active page would be destroyed (Page B)
and the previous page in history would be completely reconstructed (Page A).

> Extensions did not need to worry about what life cycle pages were in because

Author: yes

> there was only one for a tab, the Active/Visible state.

Author: fixed

> {% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/VR5hJLKud7lNIou8baHx.png", alt="Eviction of active page", width="350", height="209" %}

Author: fixed

> With Back Forward Cache and Prerendering, there is no longer a 1:1 relationship

Author: fixed

between tabs and pages. Now, each tab actually stores multiple pages and pages
transition between states rather than being destroyed and reconstructed.

For example, a page could begin its life as a prerendered (not visible) page,
transition into an active (visible) page when the user clicks a link, and then

> be stored in the Back Forward Cache (not visible) when the user navigates to

Author: fixed

another page, all without the page ever being destroyed. Later in this article
we will look at the new properties exposed to help extensions understand what
state pages are in.

> {% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/0DT7Tzx0vQhNZgegsoEj.png", alt="Types of Pages", width="531", height="209" %}

Author: fixed

> Note that a tab can have a series of prerender pages (not just one), a single

Author: fixed

*active* (visible) page, and a series of Back/Forward cached pages.

## What’s changing for extension developers?
### FrameId == 0
In Chromium, we refer to the topmost/main frame as the outermost frame.

Extension authors that assume the [*frameId*](/docs/extensions/reference/webNavigation/#a-note-about-frame-ids:~:text=onErrorOccurred%20event%20fired.-,frameId,-number)
of the outermost frame is 0 (a previous best practice) may have problems.
Since a tab can now have multiple outermost frames (prerendered and cached
pages), the assumption that there is a single outermost

> frame for a Tab is incorrect. `frameId == 0` will still continue to represent

Author: fixed

the outermost frame of the **active** page, but the outermost frames of
**other** pages in the same tab will be non-zero. A new field **frameType** has
been added to fix this problem. See the [“How do I determine if a frame is the outermost frame?”](#outermost-frame)
section of this post.

> ### Lifecycle of frame vs document

Author: fixed

Another concept that is problematic with extensions is the lifecycle of the
frame. A frame hosts a document (which is associated with a committed URL).
The document can change (say by navigating) but the *frameId* won’t, and so it
is difficult to associate that something happened in a specific document with
just *frameIds*. We are introducing a concept of a [documentId](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=retrieve%20information%20about.-,documentId,-string%C2%A0optional)

> which is a unique identifier per document. If a frame is navigated and opens a

Author: fixed

new document the identifier will change. This field is useful for determining
when pages change their lifecycle state (between prerender/active/cached)
because it remains the same.

### Web navigation events

> WebNavigation events can fire multiple times on the same page depending on the

Author: fixed

lifecycle it is in. See [“How do I tell what life cycle the page is in?”](#lifecyle)
and [“How do I determine when a page transitions?”](#transitions) sections.

## How do I tell what life cycle the page is in? {: #lifecyle}

> The [documentLifecycle](/docs/extensions/reference/extensionTypes/#type-DocumentLifecycle)

Author: fixed

> attribute has been added to a number of extensions APIs where the *frameId* was

Author: fixed

> previously available. If the value is present on an event (such as

Author: fixed

> [onCommitted](/docs/extensions/reference/webNavigation/#event-onCommitted)),

Author: fixed

> the value is the state at which the event was generated. You can always query

Author: fixed

information from the [WebNavigation GetFrame/GetAllFrames](/docs/extensions/reference/webNavigation/#method-getFrame)
APIs, but using the value from the event is always preferred. If you do use the
[GetFrame](/docs/extensions/reference/webNavigation/#method-getFrame) API be
aware the state of the frame may have changed from when the event was generated
to when the GetFrame promise is resolved.

> The [documentLifecycle](/docs/extensions/reference/extensionTypes/#type-DocumentLifecycle)

Author: fixed

## How do I determine if a frame is the outermost frame? {: #outermost-frame}

Previously extensions may have checked whether `frameId == 0` to determine
if the event occurring is for the outermost frame or not. With multiple pages
in a tab we now have multiple outermost frames, so the definition of *frameId*
is problematic. You will never receive events about a Back/Forward cached
frame. However, for prerender frames the *frameId* will be
non-zero for the outermost frame. So using `frameId == 0` as a signal for
determining if it is the outermost frame is incorrect.

To help with this, we introduced a new field called [*frameType*](/docs/extensions/reference/extensionTypes/#type-FrameType)
so determining if the frame is indeed the outermost frame is now easy.

The [*frameType*](/docs/extensions/reference/extensionTypes/#type-FrameType)
has the following values:

- `"outermost_frame"`: Typically referred to the topmost frame. Note that there are multiple of
these: there could be many for the various documentLifecycles (such as a
prerendered outermost_frame and a cached outermost_frame).
- `"fenced_frame"`: A special type of frame that is embedded inside other frames

> (see [fenced-frames](https://github.com/WICG/fenced-frame))

Reviewer: The odds are about 1.5 million to 1 that this value can even be used on a particular site. (My numbers are based on metrics for Chrome and origin trials.) I was trying to convince you to remove this completely. How about just saying 'Reserved for future use'?

- `"sub_frame"`: Typically an iframe

We can combine *documentLifecycle* with *frameType* and determine if a frame is
the active outermost frame. For example, `tab.documentLifecycle == “active” && frameType == “outermost_frame”`

## How do I solve time of use problems with frames?

As we said above a frame hosts a document and the frame may navigate to a new

> document, but the *frameId* will not change. This creates problems when you

Author: fixed

> receive an event with just a *frameId*. If you [look up the URL](/docs/extensions/reference/webNavigation/#method-getFrame)

Author: Seems to me it is here... https://developer.chrome.com/docs/extensions/reference/webNavigation/#method-getFrame:~:text=the%20main%20frame.-,url,-string

of the frame it might be different than when the event occured, this is called

> a Time of Use issue.

Author: fixed

> To address this, we introduced [*documentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=retrieve%20information%20about.-,documentId,-string%C2%A0optional)

Author: fixed

> (and [*parentDocumentId*](/docs/extensions/reference/webNavigation/#method-getFrame:~:text=parentDocumentId)).

Author: fixed

The [webNavigation.getFrame()](/docs/extensions/reference/webNavigation/#method-getFrame)

> method now makes the *frameId* optional if a *documentId* is provided. The

Author: fixed

> *documentId* will change whenever a frame is navigated.

Author: fixed.

## How do I determine when a page transitions? {: #transitions}

> There are explicit signals to determine when a page transitions these states.

Author: fixed

> Let’s look at the [WebNavigation events](/docs/extensions/reference/webNavigation/#event).

Author: fixed

For a very first navigation of any page you will see 4 events. Note that these
4 events could occur with the *documentLifecycle* state being either *prerender*

> or *active*.

Author: fixed

```js
onBeforeNavigate -> onCommitted -> onDOMContentLoaded -> onCompleted

> ```

Author: fixed

> This is illustrated in the picture below loading the prerender page with
> *documentId* == xyz.

Reviewer: ```suggestion
This is illustrated in the diagram below which shows the `documentId` changing to `"xyz"` when
the prerendered page becomes the active page.
```

Author: fixed

> {% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/p0ulwify9TGLuLh6RJpW.png", alt="ALT_TEXT_HERE", width="350", height="463" %}

Reviewer: ```suggestion

<figure>
  {% Img src="image/zKrSUSkPboWMTTSEkowJbqw5Egi2/p0ulwify9TGLuLh6RJpW.png", alt="The documentId changes when the prerendered page becomes the active page", width="350", height="463" %}
  <figcaption>The `documentId` changes when the prerendered page becomes the active page.</figcaption>
</figure>

```

Author: fixed

When a page transitions from either Back/Forward Cache or prerender to the

> active state there will be 3 more events (but with *documentLifecyle*

Author: fixed

> being *active*).

Author: fixed

```js
onBeforeNavigate -> onCommitted -> onCompleted

> ```

Author: fixed

> The *documentId* will remain the same as the original events. This is

Author: fixed

illustrated above when *documentId* == xyz activates. Note that the

> same navigation events fire, except from the ```onDOMContentLoaded```

Reviewer: ```suggestion
same navigation events fire, except for the `onDOMContentLoaded`
```

Author: fixed

> event because the page already has been loaded.

Reviewer: ```suggestion
event because the page has already been loaded.
```

Author: fixed

The [`DocumentLifecycle`](/docs/extensions/reference/extensionTypes/#type-DocumentLifecycle)
has the following values:

- `"prerender`" : Not currently presented to the user but preparing to possibly be displayed to the user.
- `"active"`: Presently displayed to the user.
- `"cached"`: Stored in the Back/Forward Cache.
- `"pending_deletion"`:  The document is being destroyed.

## How do I determine if a frame is the outermost frame? {: #outermost-frame}

Previously extensions may have checked whether `frameId == 0` to determine
if the event occurring is for the outermost frame or not. With multiple pages
in a tab we now have multiple outermost frames, so the definition of *frameId*
is problematic. You will never receive events about a Back/Forward cached
frame. However, for prerendered frames the `frameId` will be
non-zero for the outermost frame. So using `frameId == 0` as a signal for
determining if it is the outermost frame is incorrect.

To help with this, we introduced a new type called [`FrameType`](/docs/extensions/reference/extensionTypes/#type-FrameType)
so determining if the frame is indeed the outermost frame is now easy.

[`FrameType`] has the following values:

> - `"outermost_frame"`: Typically referred to the topmost frame. Note that there are multiple of
> these: there could be many for the various documentLifecycles (such as a
> prerendered outermost_frame and a cached outermost_frame).

Reviewer: ```suggestion
- `"outermost_frame"`: Typically referred to the topmost frame. Note that there are multiples of
these. For example, if you have prerendered and cached pages, each has an outermost frame that could be called its topmost frame.
```

Reviewer: The first part of the description is missing a word. Which of these is correct:

* Typically refers to the topmost frame.
* Typically referred to as the topmost frame.

Author: Done
